### DISCUSSION TITLE
useActors

### DISCUSSION THREAD

**Discussion by jtmthf** [statelyai/xstate#3275] [Ideas]
In some situations you may have a machine that has a dynamic number of actors such as the todo example given in the documentation.

```ts
import { createMachine, spawn } from 'xstate';
import { todoMachine } from './todoMachine';

const todosMachine = createMachine({
  // ...
  on: {
    'NEW_TODO.ADD': {
      actions: assign({
        todos: (context, event) => [
          ...context.todos,
          {
            todo: event.todo,
            // add a new todoMachine actor with a unique name
            ref: spawn(todoMachine, `todo-${event.id}`)
          }
        ]
      })
    }
    // ...
  }
});
```

You could subscribe to one of these individual actors in React using `useActor` but what if you wanted to subscribe to all or some of them? One solution would be to render an array of components for each actor where each has a `useActor` hook but that requires you to render a component when you may not have an associated UI and prevents from using this implementation as part of a custom hook.

## API

One solution could be to create a new `useActors` or `useActorArray` hook that subscribes to an array of actors. The API could look as below.

```ts
export function useActors<TActor extends ActorRef<any, any>>(
  actorRefs: TActor[],
  getSnapshot?: (actor: TActor) => EmittedFromActorRef<TActor>[]
): [EmittedFromActorRef<TActor>, TActor['send']];
export function useActors<TEvent extends EventObject, TEmitted>(
  actorRefs: ActorRef<TEvent, TEmitted>[],
  getSnapshot?: (actor: ActorRef<TEvent, TEmitted>) => TEmitted
): [TEmitted, Sender<TEvent>][];
```

It's intended that you're passing a homogenous array of actors, which is why this API would only takes a single `getSnapshot`.

## Usage

```ts
function useMyCustomHook() {
  const [todosState] = Machine(todosMachine);
  const todos = useActors(todosState.context.todos);

  todos.forEach(([state, send]) => {
    // do something with each actor...
  });
}
```

**Comment by davidkpiano** [comment]
This is interesting, and maybe something that can be done (for now?) in userland:

```js
// idea
const todos = useActor(aggregateActors(todosState.context.todos)); // creates a single actor
```

**Reply by polarisdigitech** [reply]
Won't this create some kind of memory leak with unsubscribed subscriptions when an actor is killed?
